_posts/2024-04-12-optimizations.md

```diff
@@ -0,0 +1,771 @@
+
+-record(bar, {bool,pid}).
+
+main(Bool) when is_boolean(Bool) ->
+    flip_state(#bar{bool=Bool,pid=self()}).
+
+flip_state(R) ->
+    R#bar{bool=not R#bar.bool}.
+```
+
+The `update_record` instruction looks like this:
+
+```
+    {update_record,{atom,inplace},
+                   3,
+                   {tr,{x,0},
+                       {t_tuple,3,true,
+                                #{1 => {t_atom,[bar]},
+                                  2 => {t_atom,[false,true]},
+                                  3 => pid}}},
+                   {x,0},
+                   {list,[2,{tr,{x,1},{t_atom,[false,true]}}]}}.
+```
+
+Based on the type for the new value, `{t_atom,[false,true]}`, the
+JIT is able to generate much shorter code than for the previous example:
+
+```
+# update_record_in_place_IsdI
+    mov rax, qword ptr [rbx]
+# skipped copy fallback because all new values are safe
+    mov rdi, qword ptr [rbx+8]
+    mov qword ptr [rax+14], rdi
+    mov qword ptr [rbx], rax
+```
+
+
+Here is the pull request for updating tuples in place:
+
+* [#8090: Destructive tuple update](https://github.com/erlang/otp/pull/8090)
+
+
+### Optimizing by generating less garbage
+
+When updating a record in place, omitting the copying of the existing
+record should be a clear win, except perhaps for very small records.
+
+What is less clear is the effect on garbage collection. Updating a
+tuple in place is an example of optimizing by generating less
+garbage. By creating less garbage, the expectation is that garbage
+collections should occur less often, which should improve the
+performance of the program.
+
+Because of the highly variable execution time for doing a garbage
+collection, it is notoriously difficult to benchmark optimizations that
+reduce the amount of garbage created. Often benchmark results don't
+generalize to doing the same operations in a real application.
+
+My own [anecdotal evidence](https://en.wikipedia.org/wiki/Anecdotal_evidence)
+suggests that in most cases there are no measurable performance wins by
+producing less garbage.
+
+I also remember when an optimization that reduced the size of an
+Erlang term resulted in a benchmark being consistently slower. It took
+the author of that optimization several days of investigation to confirm
+that the slowdown in the benchmark was not the fault of his optimization,
+but by creating less garbage, garbage collection happened at a later time
+when it happened to be much more expensive.
+
+On average we expect that this optimization should improve performance,
+especially for large records.
+
+### Optimization of funs
+
+The internal representation of funs in the runtime system has
+changed in Erlang/OTP 27, making possible several new optimizations.
+
+As an example, consider this function:
+
+
+```erlang
+madd(A, C) ->
+    fun(B) -> A * B + C end.
+```
+
+In Erlang/OTP 26, the native code for creating the fun looks like so:
+
+```
+# i_make_fun3_FStt
+L38:
+    long mov rsi, 9223372036854775807
```

Comment:
`9223372036854775807` is a placeholder value to force a long move, which might get confusing further down when we have several values in the same example. Perhaps we should explain what the values are? e.g `; fun dispatch table`

Reply:
Will do.